shared/utils/typedefs-string.hpp

```diff
@@ -102,19 +117,110 @@ struct StringW {
     constexpr StringW(std::nullptr_t npt) noexcept : inst(npt) {}
     constexpr StringW() noexcept : inst(nullptr) {}
 
-    operator std::string();
+    constexpr void const* convert() const noexcept {
+        return inst;
+    }
     constexpr void* convert() noexcept {
         return inst;
     }
+    constexpr operator Il2CppString const*() const noexcept {
+        return inst;
+    }
     constexpr operator Il2CppString*() noexcept {
         return inst;
     }
+    constexpr Il2CppString const* operator->() const noexcept {
+        return inst;
+    }
     constexpr Il2CppString* operator->() noexcept {
         return inst;
     }
-    operator std::u16string();
-    operator std::wstring();
+    constexpr operator bool() const noexcept {
+        return inst != nullptr; 
+    }
+
+    // the argument here is unused but we don't care. the comparison against nullptr more clearly shows what this operator is for imo
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wunused-parameter"
+    constexpr bool operator ==(std::nullptr_t rhs) const noexcept {
+        return inst == nullptr;
+    }
+    #pragma clang diagnostic pop
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    StringW& operator +=(const T rhs) noexcept {
+        inst = strappend(inst, rhs);
+        return *this;
+    }
+
+    StringW& operator +=(StringW rhs) noexcept {
+        inst = il2cpp_utils::detail::alloc_str(this->operator std::u16string().append(rhs));
+        return *this;
+    }
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    StringW operator +(const T rhs) const noexcept {
+        return strappend(inst, rhs);
+    }
+
+    StringW operator +(StringW rhs) const noexcept {
+        return this->operator std::u16string().append(rhs);
+    }
+
+    bool operator <(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    bool operator <(const T rhs) const noexcept {
+        return strless(inst, rhs);
+    }
+
+    template<int sz>
+    constexpr bool operator ==(ConstString<sz> const& rhs) const noexcept {
+        Il2CppChar* first = inst->chars;
+        char16_t const* second = rhs.chars;
+        Il2CppChar* firstEnd = first + inst->length;
+        char16_t const* secondEnd = second + sz;
```

Comment:
Since it seems `ConstString<sz>` is still constructed from a char array where: `sz` is deduced to be the array length, it means that: `ConstString x = "abc";` will actually have `sz = 4`. This means that in this case, where we start at the beginning of the characters and walk forward, would be incorrect, since if we have a string of `abc` in `this`, and are comparing it to the previous `ConstString`, we will count one extra.

This can be fixed in a few ways, nearly all seem to be reasonable:
1. Make the `sz` for `ConstString` actually be 1 less than the string constant length
2. Use `sz - 1` here, but also note that you would have to make sure that everywhere else where `sz` is used needs to also be aware of that.
3. Use `rhs.length`

Hopefully that gives you some ideas.